nodebuilder/node/uptime.go

```diff
@@ -0,0 +1,90 @@
+package node
+
+import (
+	"context"
+	"time"
+
+	"go.opentelemetry.io/otel/metric/global"
+	"go.opentelemetry.io/otel/metric/instrument"
+	"go.opentelemetry.io/otel/metric/instrument/asyncfloat64"
+)
+
+// OptlCollectPeriodInSeconds is the metrics collection
+// period for the OpenTelemetry callbacks in seconds.
+const OptlCollectPeriodInSeconds = 2
+
+// UptimeMetrics is a struct that records
+//
+//  1. node start time: the timestamp when the node was started
+//  2. node uptime: total time the node has been running counted in units of 1 second
+//
+// the node start time is recorded imperatively when RecordNodeStartTime is called
+// whereas the node uptime is recorded periodically
+// upon callback recalling (re-mettering from optl)
+type UptimeMetrics struct {
+	// nodeStartTS is the timestamp when the node was started.
+	nodeStartTS asyncfloat64.Gauge
+
+	// totalNodeRunTime is the total time the node has been running.
+	totalNodeRunTime asyncfloat64.Counter
+
+	// totalNodeUpTimeTicks is the total time the node has been running
+	// counted in units of `optlCollectPeriodInSeconds`
+	totalNodeUpTimeTicks int
+}
+
+var (
+	meter = global.MeterProvider().Meter("node")
+)
+
+// NewUptimeMetrics creates a new UptimeMetrics
+// and registers a callback to re-meter the totalNodeRunTime metric.
+func NewUptimeMetrics() (*UptimeMetrics, error) {
+	nodeStartTS, err := meter.
+		AsyncFloat64().
+		Gauge(
+			"node_start_ts",
+			instrument.WithDescription("timestamp when the node was started"),
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	totalNodeRunTime, err := meter.
+		AsyncFloat64().
+		Counter(
+			"node_runtime_counter_in_seconds",
+			instrument.WithDescription("total time the node has been running"),
+			instrument.WithUnit("seconds"),
+		)
+	if err != nil {
+		return nil, err
+	}
+
+	m := &UptimeMetrics{
+		nodeStartTS:          nodeStartTS,
+		totalNodeRunTime:     totalNodeRunTime,
+		totalNodeUpTimeTicks: 1,
+	}
+
+	err = meter.RegisterCallback(
+		[]instrument.Asynchronous{
+			totalNodeRunTime,
+		},
+		func(ctx context.Context) {
+			m.totalNodeUpTimeTicks += OptlCollectPeriodInSeconds
+			totalNodeRunTime.Observe(ctx, float64(m.totalNodeUpTimeTicks))
```

Comment:
consider storing last observation time and collecting seconds by using something like `time.Since(lastObservationTime).Seconds()`

Reply:
You need to increment the counter by amount of seconds passed since last observation (time.Since().Seconds()). You don't need to have any constants for that, nor modifying internals of collector.